### DISCUSSION TITLE
sqlite DNS logs on tmpfs

### DISCUSSION THREAD

**Discussion by ackstorm23** [TechnitiumSoftware/DnsServer#758] [Q&A] [answered]
I would like to poll logs using the API, which requires the sqlite database app.   However one of my main goals with polling logs is to avoid excessive wear and tear on the flash memory of the router where I run DnsServer.   

I was thinking that I could add a tmpfs  mount in the the container, where the sqlite database is written, to prevent it from causing that media wear. 

However I have two  concerns:
1. What happens if the container is restarted and the folder container the database is now empty upon startup.  Will it automatically regenerate all of the files and start the "app" with  fresh database, or will it fault and require manual intervention to start the 'app' again?
2. Is the database circular such that old records will be pruned to make way for new ones  and retain a maximum database size, it will it simply grow indefinitely?  (I know it has   "maxLogDays":  but that is not the same as a size limit, and any  day with an excessively large amount of requests could easily overflow disk storage.  This true with log files as well)

ps: it would be fantastic to have the option to run this database in-memory only so that the more advanced manual configuration of a tmpfs mount was not a blocker for less experienced users

**Comment by ShreyasZare** [comment, marked as answer]
Thanks for the post. As you mentioned, you can mount the tmpfs in the container as a volume and then set the database file path for that volume from the app config. The db file gets created if it does not exists.

The maxLogDays is the only option at the moment to remove old data. Another option can be planned to limit the number of records too. Will think on this and plan something.

What is the usecase for in-memory database that you have since any restart would mean you lose the data?

**Reply by ackstorm23** [reply]
Thank you for confirmation that it will automatically recreate and populate the database with new entries upon restart.

My use-case for in-memory database has two parts:

1. By itself, it allows for anyone to see recent DNS queries without having available disk space, concerns over disk performance or media wear, or requiring external systems to push logs to. 

2. When paired with the app, logs can be pulled in intervals from the API into another system for long-term storage.

**Reply by ShreyasZare** [reply]
That's interesting usecase and I think with small changes will be doable as sqlite supports in-memory db. Will plan this to be implemented.

**Reply by WestFarmer** [reply]
@ShreyasZare  `maxLogDays`, how to specify this ? don't see this in environment variables.

I allocated 1G PV for technitium, now service down due to no more space. I can't access web UI anymore.

**Comment by ShreyasZare** [comment]
> @ShreyasZare `maxLogDays`, how to specify this ? don't see this in environment variables.

Just switch to the Apps section on the GUI and click on Config button for the app to edit it.

> I allocated 1G PV for technitium, now service down due to no more space. I can't access web UI anymore.

You will need to delete the sqlite db file manually from disk. You can find it in the `/etc/dns/apps/Query\ Logs\ \(Sqlite\)/` folder.

**Comment by ShreyasZare** [comment]
I have released Query Logs (Sqlite) v4.1 which supports in-memory database and also includes a new option to limit the number of records the db can contain. Do update the app and let me know your feedback.

**Reply by ackstorm23** [reply]
so far, so good. :crossed_fingers:

**Reply by ShreyasZare** [reply]
Thanks for letting me know.

**Reply by SeriousHoax** [reply]
> I have released Query Logs (Sqlite) v4.1 which supports in-memory database and also includes a new option to limit the number of records the db can contain. Do update the app and let me know your feedback.

Hi @ShreyasZare! I was wondering if you will consider adding an option to save the query log on disk when the DNS server stops if in-memory database is enabled; similar to the Save Cache To Disk option.

**Reply by ShreyasZare** [reply]
@SeriousHoax Thanks for the request. It would be nice if you can explain the use-case so that the requirement is more clear. The issue with storing on disk is that when the DNS server starts, the app would again start a new in-memory db so data on disk wont be visible on the panel. If you wish to reload the data then it could create out of memory issues, crashing the DNS server.

**Reply by ZTHawk** [reply]
Even though old, the usecase to have a mix of in-memory and disk is to reduce disk IO.
- Keep database in memory
- When service is stoping then write it to disk
- When service starts then load file from disk into memory

I think this would improve lifetime of SSDs.

I have found this:
- https://sqlite.org/c3ref/serialize.html
- https://sqlite.org/c3ref/deserialize.html
- https://stackoverflow.com/a/78789206

**Reply by ZTHawk** [reply]
I have just tested those functions in a small test application and they seem to work. I can:
- deserialize from file into memory
- serialize from memory into file

@ShreyasZare I will create a PR for that (with additional settings parameters):
- inMemoryDbSyncToFile (bool) (allows to serialize/deserialize)
- inMemoryDbSyncInterval (integer in minutes) (automatic serialization every X minutes; mitigate data loss on potential app crash)

My intention ist to reduce IO. Currently I have ~250 queries per minute but the IO is ~1.2MB per minute. I do not understand why 250queries could create such high IO. This is ~4KB per query.

**Reply by ShreyasZare** [reply]
@ZTHawk Thanks for the response. If you have working code for this then do create a PR. I would suggest that you test it on both Windows and Linux platform just in case if it needs platform specific changes.

> My intention ist to reduce IO. Currently I have ~250 queries per minute but the IO is ~1.2MB per minute. I do not understand why 250queries could create such high IO. This is ~4KB per query.

It could be due to 4KB block size for the file system.

**Reply by ZTHawk** [reply]
Deserializing a database switches it to private (not shared anymore). This makes any new created connection to use a different in memory DB. Luckily there is a workaround.
The workaround opens the database in memory as private and then copies it over to the shared one.

Options:
- Use copy over workaround
  - How big could the databases be? (increased memory usage during the deserialization)
  - I think I go for this solution as it is the least change to the code.
- Change to a single sqlite connection object (either for in memory only or both)
  - Cleaner code
  - No need to do the workaround (copy DB over)
  - Requires more code change
  - Are there any known issues (regarding threads) when using only one connection?
